presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java

```diff
@@ -93,7 +93,7 @@ public RuleAssert on(Function<PlanBuilder, PlanNode> planProvider)
         return this;
     }
 
-    public void doesNotFire()
+    public RuleAssert doesNotFire()
```

Comment:
I don't see benefit. Rule & input is set up, so the expected output is quite well defined. What's the use case?

Reply:
Consider that now you are able:

```
assertThat(rule)
   .on(somePlan)
   .matches(pattern_for_feature1)
   .matches(pattern_for_feature2);
```

or


```
assertThat(rule)
   .on(somePlan)
   .matches(pattern)
   .on(plan2)
   .doesNotFire();
```

Especially it will be needed to test plane shape together with its traits.

Reply:
well, but tests for features 1,2 will usually land in different test method. To re-use `somePlan` you can always extract it to a builder method (see usages of `com.facebook.presto.sql.planner.iterative.rule.TestEliminateCrossJoins#crossJoinAndJoin`).

I am OK with the change, but I just don't feel convinced

Reply:
+1 @findepi 
There is not much benefit in:
```
assertThat(rule)
   .on(somePlan)
   .matches(pattern)
   .on(plan2)
   .doesNotFire();
```
but it could make harder to debug tests.

Reply:
> well, but tests for features 1,2 will usually land in different test method. 

I can agree in regards of second example, but in first when check two features it might be more fluent to have it in a chain, not as separate method. I would prefer to not force user, but to give them possibility to choose tools which fit the best for their situations.

> but it could make harder to debug tests (e.g: spock and stats)

I do not understand why, each invocation is using a separate `Memo` instance and `RuleApplication`.

What about matching plan and traits together?
```
assertThat(rule)
   .on(somePlan)
   .matches(pattern)
   .hasTrait(trait1)
   .satisfiesTrait(trait2);
```

Reply:
The last example (matching plan pattern and traits) is most convincing but again -- this would not be a rule test. A rule either produces a new node (hence: match pattern) or applies a trait to existing one (hence: `hasTrait`). Maybe let's leave this change until we see evidence it's really useful?